test/oph/ehoks/hoks/hoks_save_test.clj

```diff
@@ -569,9 +572,110 @@
+              #{["ei_laheteta" "aloittaneet"]
+                ["ei_laheteta" "valmistuneet"]})
+          (eq (set (map (juxt :kyselytyyppi :syy)
+                        (tapahtumat/get-all-by-hoks-id-and-kyselytyypit!
+                          db/spec
+                          {:hoks-id (:id hoks-db)
+                           :kyselytyypit ["aloittaneet" "valmistuneet"]})))
+              #{["valmistuneet" "hoks_tallennettu"]
+                ["valmistuneet" "ei_ole"]
+                ["aloittaneet" "hoks_tallennettu"]
+                ["aloittaneet" "ei_ole"]})
+          ;; with changed dates in HOKS
+          (hoks/replace!
+            (:id hoks-db)
+            (-> (:id hoks-db)
+                (hoks/get-by-id)
+                (test-utils/dissoc-module-ids)
+                (assoc :osaamisen-saavuttamisen-pvm (LocalDate/of 2022 11 1))
+                (assoc :osaamisen-hankkimisen-tarve true)
+                (update
+                  :hankittavat-paikalliset-tutkinnon-osat
+                  (fn [hptos]
+                    (map (fn [hpto]
+                           (update
+                             hpto :osaamisen-hankkimistavat
+                             (fn [ohts]
+                               (map #(assoc % :keskeytymisajanjaksot
+                                            [{:alku (LocalDate/of 2019 2 15)}])
+                                    ohts))))
+                         hptos)))
+                (update
+                  :hankittavat-ammat-tutkinnon-osat
+                  (fn [hatos]
+                    (map (fn [hato]
+                           (update
+                             hato :osaamisen-hankkimistavat
+                             (fn [ohts]
+                               (map #(assoc % :alku (LocalDate/of 2019 05 15)
+                                            :loppu (LocalDate/of 2019 07 15))
+                                    ohts))))
+                         hatos)))))
+          (eq (set (map (juxt :tila :kyselytyyppi :jakson-yksiloiva-tunniste
+                              :voimassa-alkupvm)
+                        (palaute/get-by-hoks-id-and-kyselytyypit!
+                          db/spec
+                          {:hoks-id (:id hoks-db)
+                           :kyselytyypit ["aloittaneet" "valmistuneet"
+                                          "tyopaikkajakson_suorittaneet"]})))
+              #{["ei_laheteta" "tyopaikkajakson_suorittaneet"
+                 "abcd" (LocalDate/parse "2019-02-16")]
+                ["odottaa_kasittelya" "tyopaikkajakson_suorittaneet"
+                 "1234567890" (LocalDate/parse "2019-07-16")]
+                ["ei_laheteta" "tyopaikkajakson_suorittaneet"
+                 "1" (LocalDate/parse "2019-03-01")]
+                ["odottaa_kasittelya" "valmistuneet"
+                 nil (LocalDate/parse "2022-11-01")]
+                ["odottaa_kasittelya" "aloittaneet"
+                 nil (LocalDate/parse "2019-03-18")]}))))))
```

Comment:
Toive: Odotettujen arvojen määrittely joko `def`illä tai `let`illä testin alussa, jolloin testin control flowta on hieman helpompi seurata. Ei ole pakko korjata tässä.

Toive: Tää testi testaa nyt jonkin verran muutakin kuin opiskelijapalautteen muodostumista HOKSin ensikertaisen tallentamisen yhteydessä. Jonkin verran overlappia on myös testien `do-not-form-opiskelijapalaute-in-hoks-update` ja `form-opiskelijapalaute-in-hoks-replace` kanssa. Tästäkään ei sinänsä oo mitään haittaa (parempi niin että testataan "liikaa" vaikka vähän toisteisesti kuin liian vähän), mutta selkeyden kannalta voisi ehkä tehdä toiminnallisuuden testaamiselle kokonaan uuden testin tai sitten pilkkoa osiin ja täydentää noita olemassa olevia testejä. Järkeen kyllä käy tehdä näitä asioita samassa testissä, koska HOKSin päivittäminen edellyttää, että HOKS on ensin tallennettu jne. Ei myöskään tarvitse korjata tässä.

Reply:
Kiitos kommenteista :)

Tuon jälkemmäisen voisin toteuttaa siten, että teen tästä oman testinsä, mutta koska sen alkuosa olisi sitten päällekkäinen HOKSin tallennustestin kanssa, otan sen pois.  Yhdessä nää tarkoittaa sitä, että oikeastaan vain testin nimi vaihtuu.

Ekan voin toteuttaa, joskus on helpompaa että control flow on hyvin näkyvissä, joskus taas että odotetut arvot ovat hyvin näkyvissä.  Ainakin tuossa toivomassasi tavassa voi hypätä helposti tarvittaessa niihin odotettuihin arvoihin ja niiden nimeäminen saattaa tehdä testistä helppolukuisemman.